docs/introduction/upgrade-to-v2.md

```diff
@@ -0,0 +1,142 @@
+---
+title: Upgrade to fp-ts 2.x
+parent: Introduction
+nav_order: 5
+has_toc: false
+---
+
+# Upgrade to version 2.x
+
+`fp-ts@2.x` brings with it some major improvements, but also breaking changes and the removal of deprecated APIs. This document will help you understand what changed and how you can upgrade your existing codebase.
+{: .fs-6 .fw-300 }
+
+---
+
+The major changes in `fp-ts@2.x` are:
+
+- Requires TypeScript 3.5+
+- `fp-ts@1.19.x` has been released with backported 2.x features for a gradual upgrade path
+- Data types are no longer implemented as classes, resulting in a new API using `pipe`
+- The `run()` method on `IO`, `Task`, etc. has been replaced with a thunk
+- Deprecations
+  - `HKT`: Replaced `Type<n>` with `Kind<n>`
+  - Replaced `Setoid` with `Eq`
+  - Several modules were removed, e.g. `Exception`, `Free`, `StrMap`, `Trace`, `Validation`, …
+  - Read the [full changelog](https://github.com/gcanti/fp-ts/pull/881) for all the changes
+
+## Upgrading from version 1.x
+
+You can gradually upgrade your existing codebase using the `fp-ts@1.19.x` release; the new `fp-ts@2.x` APIs have been backported to this release.
+
+1. Upgrade TypeScript to version 3.5+
+1. Install `fp-ts@1.19.x`, which contains the new `fp-ts@2.x` APIs
+1. Optional: activate the `@obsolete` rule for `tslint` to get guidance on what to change
+1. Familiarise yourself with [the new API](https://github.com/gcanti/fp-ts/pull/881)
+1. Gradually replace the existing code with the new API
+1. Upgrade to `fp-ts@2.x` and make sure to also upgrade all dependencies that rely on `fp-ts`
+
+### tslint rule
+
+In order to make easier to spot all the occurrences of chainable APIs without depending on `@deprecated`, which would force you to migrate in one shot, a custom tslint rule is provided (`@obsolete`).
+
+Add the following lines to your `tslint.json` to turn the `@obsolete` rule on:
+
+```diff
+{
++  "rulesDirectory": ["./node_modules/fp-ts/rules"],
+   "rules": {
++    "obsolete": true
+   }
+}
+```
+
+### Dependencies
+
+Make sure to always have a single version of `fp-ts` installed in your project. Multiple versions are known to cause `tsc` to hang during compilation. You can check the versions currently installed using `npm ls fp-ts`: make sure there's a single version and all the others are marked as `deduped`.
```

Comment:
I think this (tsc hanging) will be no longer the case with fp-ts >=2 ? I remember someone mentioning it in multiple occasions, but never verified myself. cc @gcanti

Reply:
If _all_ the versions you have installed at >= 2 then I _think_ the issue will no longer occur.

Reply:
I think you're right, this was one of the motivations behind the removal of classes.